### PR TITLE
Fix research sync issues

### PR DESCRIPTION
**Pull request by pGlase**
PR aims to adress the following issues, which were taken from Baldie's notes.
This is also a mindlist of the current state.

- [x] Occasional desync of the client's research-queues compared to the host.
       Note: This seems to happen when some research finishes.
       State: I think it's linked to the next point.
- [x] Occasionally, clients see research as completed, even though it is not completed.
       State: Added statement that clients cannot unlock tech to prevent clients unlocking prematurely.
       Progress: Honestly, I've seen (too) many research enqueueing, cancelling and finishing and never saw this happening again after preventing clients from ever unlocking tech with their own research.

- [x] If the research consumes items from inventory and current research is changed, occasionally, your items are not refunded.
Fix: Host now keeps track of research (hashes) added by clients and issues refund packets on cancelling.

- [?] Research will only progress if at least one player has all nessecary items in their inventory. 
Player A enqueues Mecha Core, but has only Copper ingots (or no items). Only Player B has Iron Plates.
->Both players loose their items, but research is not starting, since Host has not yet started research (missing iron ingots). See 1st picture, minor issue due to seperated research inventories.
grafik
Since we don't have shared inventories, this seems honestly a bit daunting regarding syncing for a minor concern.
  
- [?] The research progresses too fast. It was consuming cubes with much higher speed that it should.
   State: Small tests with Fluid Storage encapsulation.
Only Host => 60hashes/min:
9000 hashes /  60 hashes/s = 150s; measured ~150s
Host + 1x Client => 120hashes/min:
9000 hashes / 120 hashes/s =  75s;measured ~75s
Host + 1x Lab => 120hashes/min:
9000 hashes / 120 hashes/s =  75s; measured ~75s
Host + 1x Client + 1x Lab: + Lab => 180hashes/min
9000 hashes / 180 hashes/s =  50s; measured ~50s
Result: Seems to work on my machine, maybe I need another research.

- [x] Improvement: Sending regular updates (maybe every 2 sec) to sync number of consumed cubes. So all players will see the same number in the research UI.
- [x] Improvement: Syncing of required cubes for the current research added.

### CODE REVIEW

File: NebulaClient/PacketProcessors/GameHistory/GameHistoryRemoveTechProcessor.cs [hunk 1]
```diff
@@ -12,10 +12,29 @@ class GameHistoryRemoveTechProcessor : IPacketProcessor<GameHistoryRemoveTechPac
...
+                if (index < 0)
+                {
+                    //sanity: packet wanted to remove tech, which is not queued on this client
+                    Log.Info($"ProcessPacket: TechId: {packet.techId} was not in queue, discarding paket");
+                    return;
+                }
+                //recover spend items from inventory before removing it from queue
+                GameMain.mainPlayer.mecha.lab.ManageTakeback();
+                GameMain.history.RemoveTechInQueue(index);
+                //Wenn nodes noch da sind: UIResearchQueue::UpdateNodes()
```
Comment (Baldie-dev) [inline NebulaClient/PacketProcessors/GameHistory/GameHistoryRemoveTechProcessor.cs:36]: Maybe this comment to english? :D

File: NebulaModel/Packets/GameHistory/GameHistoryRemoveTechPacket.cs [hunk 2]
```diff
@@ -2,13 +2,13 @@
 {
     public class GameHistoryRemoveTechPacket
     {
-        public int Index { get; set; }
+        public int techId { get; set; }
```
Comment (Baldie-dev) [inline NebulaModel/Packets/GameHistory/GameHistoryRemoveTechPacket.cs:5]: There is naming convention to have first letter capital for the class properties :)

File: NebulaHost/PacketProcessors/GameHistory/GameHistoryResearchContributionProcessor.cs [hunk 3]
```diff
@@ -13,7 +14,11 @@ public void ProcessPacket(GameHistoryResearchContributionPacket packet, NebulaCo
             //Check if client is contributing to the correct Tech Research
             if (packet.TechId == GameMain.history.currentTech)
             {
+                Log.Info($"ProcessPacket researchContribution: got package for same tech");
                 GameMain.history.AddTechHash(packet.Hashes);
+                PlayerManager playerManager = MultiplayerHostSession.Instance.PlayerManager;
+                playerManager.GetPlayer(conn).UpdateResearchProgress(packet.TechId, packet.Hashes);
+                Log.Info($"ProcessPacket researchContribution: playerid by: {playerManager.GetPlayer(conn).Id} - hashes {packet.Hashes}");
```
Comment (Mushroom) [inline NebulaHost/PacketProcessors/GameHistory/GameHistoryResearchContributionProcessor.cs:21]: Do we maybe want to make this a debug log instead? Could get spammy

File: NebulaPatcher/Patches/Dynamic/GameHistoryData_Patch.cs [hunk 4]
```diff
@@ -141,5 +141,26 @@ public static bool Prefix5()
...
+        public static bool Prefix6()
+        {
+            //Wait for the authoritative packet for unlocking tech features in multiplayer for clients
+            return !SimulatedWorld.Initialized || LocalPlayer.IsMasterClient || GameDataHistoryManager.IsIncomingRequest;
+        }
+
+        [HarmonyPrefix]
+        [HarmonyPatch("RemoveTechInQueue")]
+        public static void Prefix7(int index, out int __state)
+        {
```
Comment (Mushroom) [inline NebulaPatcher/Patches/Dynamic/GameHistoryData_Patch.cs:156]: Is there a missed singleplayer chekc here?

File: NebulaHost/PacketProcessors/GameHistory/GameHistoryResearchContributionProcessor.cs (hunk 3 above)
Comment (pGlase) [reply NebulaHost/PacketProcessors/GameHistory/GameHistoryResearchContributionProcessor.cs:21]: Good catch, I missed that one

File: NebulaHost/PacketProcessors/GameHistory/GameHistoryRemoveTechProcessor.cs [hunk 5]
```diff
@@ -3,6 +3,8 @@
 using NebulaModel.Packets.GameHistory;
 using NebulaModel.Packets.Processors;
 using NebulaWorld.GameDataHistory;
+using NebulaModel.Logger;
```
Comment (Baldie-dev) [inline NebulaHost/PacketProcessors/GameHistory/GameHistoryRemoveTechProcessor.cs:6]: Logger is not used in this class, right?

File: NebulaPatcher/Patches/Dynamic/GameHistoryData_Patch.cs (hunk 4 above)
Comment (pGlase) [reply NebulaPatcher/Patches/Dynamic/GameHistoryData_Patch.cs:156]: I've added if (**SimulatedWorld.Initialized &&** LocalPlayer.IsMasterClient), does this address your concern or do i need to add something else?

File: NebulaHost/PacketProcessors/GameHistory/GameHistoryTechRefundProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        public void ProcessPacket(GameHistoryTechRefundPacket packet, NebulaConnection conn)
+        {
+            //only refund if we have contributed
+            if(packet.TechHashedContributed > 0)
+            {
+                //client should have the same research queued, seek currently needed itemIds and re-add points that were contributed
+                foreach (KeyValuePair<int, int> item in GameMain.data.mainPlayer.mecha.lab.itemPoints.items)
+                {
+                    GameMain.data.mainPlayer.mecha.lab.itemPoints.Alter(item.Key, (int)packet.TechHashedContributed * 3600);
```
Comment (Baldie-dev) [inline NebulaHost/PacketProcessors/GameHistory/GameHistoryTechRefundProcessor.cs:21]: Can you replace it please with single reference:
```cs
ItemPack itemPoints = GameMain.data.mainPlayer.mecha.lab.itemPoints;
```

and then have it in the loop:
```cs
foreach (KeyValuePair<int, int> item in itemPoints)
...
itemPoints.Alter(item.Key, (int)packet.TechHashedContributed * 3600);
```

File: NebulaHost/Player.cs [hunk 7]
```diff
@@ -1,5 +1,6 @@
 ﻿using NebulaModel.DataStructures;
 using NebulaModel.Networking;
+using NebulaModel.Logger;
```
Comment (Baldie-dev) [inline NebulaHost/Player.cs:3]: Logger is not used, right?

Comment (pGlase) [reply NebulaHost/Player.cs]: Removed

File: NebulaHost/PacketProcessors/GameHistory/GameHistoryRemoveTechProcessor.cs (hunk 5 above)
Comment (pGlase) [reply NebulaHost/PacketProcessors/GameHistory/GameHistoryRemoveTechProcessor.cs]: Removed

File: NebulaHost/PacketProcessors/GameHistory/GameHistoryTechRefundProcessor.cs (hunk 6 above)
Comment (pGlase) [reply NebulaHost/PacketProcessors/GameHistory/GameHistoryTechRefundProcessor.cs]: Good catch :)

**Review by Baldie-dev — APPROVED** [review]

**Comment by Baldie-dev** [comment]
Btw. this is linked to issue #183

**Comment by Mushroom** [comment]
@pGlase Please fix release build issues

File: NebulaHost/PacketProcessors/GameHistory/GameHistoryTechRefundProcessor.cs [hunk 8]
```diff
@@ -16,9 +15,10 @@ public void ProcessPacket(GameHistoryTechRefundPacket packet, NebulaConnection c
             if(packet.TechHashedContributed > 0)
             {
                 //client should have the same research queued, seek currently needed itemIds and re-add points that were contributed
-                foreach (KeyValuePair<int, int> item in GameMain.data.mainPlayer.mecha.lab.itemPoints.items)
+                ItemPack itemPoints = GameMain.data.mainPlayer.mecha.lab.itemPoints;
+                foreach (KeyValuePair<int, int> item in itemPoints)
```
Comment (Baldie-dev) [inline NebulaHost/PacketProcessors/GameHistory/GameHistoryTechRefundProcessor.cs:19]: I think it should be:
```cs
foreach (KeyValuePair<int, int> item in itemPoints.items)
```

**Review by Baldie-dev — CHANGES_REQUESTED** [review]

**Comment by pGlase** [comment]
Sorry to react slow, I had a busy day at work, will fix it now.

**Review by Baldie-dev — APPROVED** [review]

**Review by Mushroom — APPROVED** [review]

**Pull request merged**